src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -994,21 +1001,27 @@ else if (def.type.isCollection())
         }
     }
 
+    private boolean needsToSkipUserLimit()
+    {
+        // if post query ordering is required, and it's not ANN
+        return needsPostQueryOrdering() && (orderingComparator == null || !orderingComparator.isAnn());
+    }
+
     private boolean needsPostQueryOrdering()
     {
-        // We need post-query ordering only for queries with IN on the partition key and an ORDER BY.
-        return restrictions.keyIsInRelation() && !parameters.orderings.isEmpty();
+        // We need post-query ordering only for queries with IN on the partition key and an ORDER BY or ANN
+        return restrictions.keyIsInRelation() && !parameters.orderings.isEmpty() || (orderingComparator != null && orderingComparator.isAnn());
```

Comment:
nit: Could we break out the `orderingComparator` test to a `needsANNOrdering` method and use it here and above?

Reply:
https://github.com/datastax/cassandra/pull/626/commits/332551b7652d4db5dde349a08439d3536c50a653